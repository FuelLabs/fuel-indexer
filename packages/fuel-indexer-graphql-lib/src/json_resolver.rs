use std::collections::HashMap;

pub use crate::spec::*;

pub(self) mod prelude {
    pub use crate::spec::*;
    pub use async_graphql::dynamic::*;
    pub use serde_json::Value as JsonValue;
}

use prelude::*;

pub struct Node(pub String, pub JsonValue);
impl Node {
    pub fn id(&self) -> &str {
        &self.0
    }
    pub fn type_id(&self) -> &str {
        self.id().split(':').next().unwrap()
    }
    pub fn local_id(&self) -> &str {
        self.id().split(':').nth(1).unwrap()
    }
    pub fn data(&self) -> &JsonValue {
        &self.1
    }

    pub fn to_field_value(&self) -> FieldValue {
        FieldValue::borrowed_any(self)
    }
    pub fn to_typed_field_value(&self) -> FieldValue {
        let type_id = self.type_id().to_string();
        FieldValue::borrowed_any(self).with_type(type_id)
    }
}

pub struct JsonResolver {
    pub nodes: HashMap<String, Node>,
}
impl JsonResolver {
    pub fn new(nodes: Vec<Node>) -> Self {
        let nodes = nodes
            .into_iter()
            .map(|node| (node.id().to_string(), node))
            .collect::<HashMap<_, _>>();
        Self { nodes }
    }
}

impl QueryResolver for JsonResolver {
    fn resolve_node_by_id(ctx: ResolverContext) -> FieldFuture {
        FieldFuture::new(async move {
            let resolver = ctx.data_unchecked::<JsonResolver>();
            let id = ctx.args.get("id").unwrap();
            let id = id.string().unwrap();
            let node = resolver.nodes.get(id);
            if let Some(node) = node {
                Ok(Some(node.to_typed_field_value()))
            } else {
                Ok(None)
            }
        })
    }
    fn resolve_node_by_ids(ctx: ResolverContext) -> FieldFuture {
        FieldFuture::new(async move {
            let resolver = ctx.data_unchecked::<JsonResolver>();
            let ids = ctx.args.get("ids").unwrap();
            let ids = ids.list().unwrap();
            let ids = ids
                .iter()
                .map(|id| id.string().unwrap().to_string())
                .collect::<Vec<_>>();
            let nodes = ids
                .iter()
                .map(|id| {
                    let node = resolver.nodes.get(id).unwrap();
                    node.to_typed_field_value()
                })
                .collect::<Vec<_>>();
            Ok(Some(FieldValue::list(nodes)))
        })
    }
}

impl NodeResolver for JsonResolver {
    fn resolve_id(
        ctx: async_graphql::dynamic::ResolverContext,
    ) -> async_graphql::dynamic::FieldFuture {
        FieldFuture::new(async move {
            let parent = ctx.parent_value.try_downcast_ref::<Node>()?;
            Ok(Some(FieldValue::value(parent.id())))
        })
    }
    fn resolve_data(ctx: ResolverContext) -> FieldFuture {
        FieldFuture::new(async move {
            let parent = ctx.parent_value.try_downcast_ref::<Node>()?;
            let path_node = ctx.path_node.unwrap();
            let field_name = path_node.field_name();
            let value = parent.data()[field_name].as_str().unwrap();
            Ok(Some(FieldValue::value(value)))
        })
    }
    fn resolve_ref(ctx: ResolverContext) -> FieldFuture {
        FieldFuture::new(async move {
            let parent = ctx.parent_value.try_downcast_ref::<Node>()?;
            let path_node = ctx.path_node.unwrap();
            let field_name = path_node.field_name();
            let id = parent.data()[field_name].as_str().unwrap();
            let resolver = ctx.data_unchecked::<JsonResolver>();
            let node = resolver.nodes.get(id).unwrap();
            Ok(Some(node.to_field_value()))
        })
    }
    fn resolve_connection(_ctx: ResolverContext) -> FieldFuture {
        unimplemented!()
    }
}

impl ConnectionResolver for JsonResolver {
    fn resolve_total_count(_ctx: ResolverContext) -> FieldFuture {
        unimplemented!()
    }
    fn resolve_nodes(_ctx: ResolverContext) -> FieldFuture {
        unimplemented!()
    }
    fn resolve_edges(_ctx: ResolverContext) -> FieldFuture {
        unimplemented!()
    }
    fn resolve_page_info(_ctx: ResolverContext) -> FieldFuture {
        unimplemented!()
    }
}

impl EdgeResolver for JsonResolver {
    fn resolve_node(_ctx: ResolverContext) -> FieldFuture {
        unimplemented!()
    }
    fn resolve_cursor(_ctx: ResolverContext) -> FieldFuture {
        unimplemented!()
    }
}

impl PageInfoResolver for JsonResolver {
    fn resolve_has_next_page(
        ctx: async_graphql::dynamic::ResolverContext,
    ) -> async_graphql::dynamic::FieldFuture {
        FieldFuture::new(async move {
            let parent = ctx.parent_value.try_downcast_ref::<JsonValue>()?;
            Ok(Some(FieldValue::value(
                parent["has_next_page"].as_bool().unwrap(),
            )))
        })
    }
    fn resolve_has_previous_page(ctx: ResolverContext) -> FieldFuture {
        FieldFuture::new(async move {
            let parent = ctx.parent_value.try_downcast_ref::<JsonValue>()?;
            Ok(Some(FieldValue::value(
                parent["has_previous_page"].as_bool().unwrap(),
            )))
        })
    }
    fn resolve_start_cursor(ctx: ResolverContext) -> FieldFuture {
        FieldFuture::new(async move {
            let parent = ctx.parent_value.try_downcast_ref::<JsonValue>()?;
            Ok(Some(FieldValue::value(
                parent["start_cursor"].as_str().unwrap(),
            )))
        })
    }
    fn resolve_end_cursor(ctx: ResolverContext) -> FieldFuture {
        FieldFuture::new(async move {
            let parent = ctx.parent_value.try_downcast_ref::<JsonValue>()?;
            Ok(Some(FieldValue::value(
                parent["end_cursor"].as_str().unwrap(),
            )))
        })
    }
}
