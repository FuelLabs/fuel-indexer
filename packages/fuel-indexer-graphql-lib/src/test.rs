pub mod prelude {
    pub use super::util::*;
    pub use assert_matches::*;
    pub use async_graphql::dynamic::*;
    pub use async_graphql::{Request, Response, ServerError};
    pub use async_trait::async_trait;
    pub use extension_trait::extension_trait;
    pub use graphql_parser::*;
    pub use insta::*;
    pub use serde_json::{json, Value as JsonValue};
    pub use velcro::hash_map;
}

pub mod util {
    pub use super::prelude::*;

    pub async fn execute_query(
        schema: &Schema,
        query: impl Into<String>,
        root_value: Option<FieldValue<'static>>,
    ) -> Result<Response, Vec<ServerError>> {
        let request = Request::new(query);
        let request = if let Some(root_value) = root_value {
            request.root_value(root_value)
        } else {
            request.into()
        };
        let response = schema.execute(request).await;
        response.into_result()
    }

    #[extension_trait]
    pub impl TestSchema for Schema {
        fn pretty_sdl(&self) -> String {
            let sdl = self.sdl();
            let ugly = parse_schema::<&str>(&sdl).unwrap();
            ugly.format(&Style::default())
        }
    }
}
