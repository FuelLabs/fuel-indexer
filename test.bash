#!/bin/bash
#
#
foo=$1

if [[ ${foo} =~ ^refs/heads/v[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
    echo "is_release_branch=true"
fi
