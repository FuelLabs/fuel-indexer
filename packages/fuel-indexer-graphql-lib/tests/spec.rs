mod test;

use crate::test::prelude::*;
use fuel_indexer_graphql_lib::json_resolver::*;

fn build_schema() -> SchemaBuilder {
    let mut schema = Schema::build(TypeRef::QUERY, None, None)
        .register(<Object as PageInfoObject<JsonResolver>>::new_page_info())
        .register(<Interface as NodeInterface>::new_node());
    let query = <Object as QueryObject<JsonResolver>>::new_query();

    // Some nodes
    let chain = <Object as NodeObject<JsonResolver>>::new_node("Chain");
    let mut block = NodeObject::<JsonResolver>::new_node("Block");
    block = <Object as NodeObject<JsonResolver>>::data_field(
        block,
        "number",
        TypeRef::named_nn(TypeRef::INT),
    );
    block = <Object as NodeObject<JsonResolver>>::data_field(
        block,
        "hash",
        TypeRef::named_nn(TypeRef::STRING),
    );
    block = <Object as NodeObject<JsonResolver>>::ref_field(
        block,
        "chain",
        TypeRef::named_nn(TypeRef::node("Chain")),
    );
    let mut transaction = <Object as NodeObject<JsonResolver>>::new_node("Transaction");
    transaction = <Object as NodeObject<JsonResolver>>::data_field(
        transaction,
        "index",
        TypeRef::named_nn(TypeRef::INT),
    );
    transaction = <Object as NodeObject<JsonResolver>>::data_field(
        transaction,
        "hash",
        TypeRef::named_nn(TypeRef::STRING),
    );

    schema = schema
        .register(chain)
        .register(block)
        .register(transaction)
        .register(query);

    schema
}

#[tokio::test]
async fn resolve() {
    // Build the schema
    let mut schema = build_schema();
    let resolver = JsonResolver::new(vec![
        Node(
            "Chain:0".to_string(),
            json!({
                "type": "Chain",
                "id": "Chain:0",
                "dummy": "0x00",
            }),
        ),
        Node(
            "Block:0".to_string(),
            json!({
                "type": "Block",
                "id": "Block:0",
                "number": "0x00",
                "hash": "0x00",
                "chainId": "0",
                "chain": "Chain:0",
            }),
        ),
    ]);
    schema = schema.data(resolver);
    let schema = schema.finish();
    assert_matches!(schema, Ok(_));
    let schema = schema.unwrap();

    // Execute a query
    let response = execute_query(
        &schema,
        r#"
          query {
              node(id: "Block:0") {
                  id
                  ... on Block {
                    hash
                    chain {
                        id
                    }
                  }
              }
          }
      "#,
        None,
    )
    .await
    .unwrap();

    assert_json_snapshot!(response.data, @r###"
    {
      "node": {
        "id": "Block:0",
        "hash": "0x00",
        "chain": {
          "id": "Chain:0"
        }
      }
    }
    "###
    );
}

#[tokio::test]
async fn sdl() {
    // Build the schema
    let schema = build_schema();
    let schema = schema.finish();
    assert_matches!(schema, Ok(_));
    let schema = schema.unwrap();

    // Print the schema
    let text = schema.pretty_sdl();
    assert_snapshot!(text);
}
