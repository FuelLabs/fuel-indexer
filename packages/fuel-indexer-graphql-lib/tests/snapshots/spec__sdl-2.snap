---
source: packages/fuel-indexer-graphql-lib/tests/spec.rs
expression: text
---
type Block implements Node {
  id: ID!
  number: Int!
  hash: String!
}

interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
  endCursor: String
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]!
}

type Transaction implements Node {
  id: ID!
  index: Int!
  hash: String!
}

schema {
  query: Query
}

